### PR TITLE
subsystems and commands for winch and hook on hanger.

### PR DESCRIPTION
**Pull request by tfrancisco**
@aschokking

### CODE REVIEW

File: CompetitionTemplate/src/competition/subsystems/hanger/HookSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package competition.subsystems.hanger;
+
+import xbot.common.command.BaseSubsystem;
+import xbot.common.controls.actuators.XSpeedController;
+import xbot.common.injection.wpi_factories.WPIFactory;
+
+public class HookSubsystem extends BaseSubsystem{
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/HookSubsystem.java]: Needs @Singleton and @Inject on constructor

File: CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package competition.subsystems.hanger;
+
+import xbot.common.command.BaseSubsystem;
+import xbot.common.controls.actuators.XSpeedController;
+import xbot.common.injection.wpi_factories.WPIFactory;
+
+public class WinchSubsystem extends BaseSubsystem{
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java]: Same thing here

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookExtendCommand.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import competition.subsystems.hanger.HookSubsystem;
+import xbot.common.command.BaseCommand;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+
+public class HookExtendCommand extends BaseCommand{
+    public HookSubsystem hookSubsystem;
+    DoubleProperty hookExtentionPower;
+    
+    public HookExtendCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookExtendCommand.java]: Needs `@Inject`

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookRetractCommand.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import competition.subsystems.hanger.HookSubsystem;
+import xbot.common.command.BaseCommand;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+
+public class HookRetractCommand extends BaseCommand{
+    public HookSubsystem hookSubsystem;
+    DoubleProperty hookRetractionPower;
+    
+    public HookRetractCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookRetractCommand.java]: `@Inject`

**Comment by aschokking** [comment]
If you added a few simple unit tests it would have caught the missing `@Inject`s ;)

**Comment by JohnGilb** [comment]
just address alex's comments and you're good to go

**Comment by aschokking** [comment]
@tfrancisco : waiting for your fixes here, I think you mentioned yesterday you had addressed the feedback but maybe the fixes weren't pushed yet?

File: CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import xbot.common.command.BaseSubsystem;
+import xbot.common.controls.actuators.XSpeedController;
+import xbot.common.injection.wpi_factories.WPIFactory;
+
+@Singleton
+public class WinchSubsystem extends BaseSubsystem{
+public XSpeedController winchMotor;
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java]: tab this line in

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookRetractCommand.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import xbot.common.properties.XPropertyManager;
+
+public class HookRetractCommand extends BaseCommand{
+    public HookSubsystem hookSubsystem;
+    DoubleProperty hookRetractionPower;
+    
+    @Inject
+    public HookRetractCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
+        this.hookSubsystem = hookSubsystem;
+        hookRetractionPower = propManager.createPersistentProperty("hook retraction power", 1.0);
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookRetractCommand.java]: this needs to require the subsystem

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookExtendCommand.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import xbot.common.properties.XPropertyManager;
+
+public class HookExtendCommand extends BaseCommand{
+    public HookSubsystem hookSubsystem;
+    DoubleProperty hookExtentionPower;
+    
+    @Inject
+    public HookExtendCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
+        this.hookSubsystem = hookSubsystem;
+        hookExtentionPower = propManager.createPersistentProperty("hook extention power", -1.0);
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/copy/HookExtendCommand.java]: missing requires

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchExtendCommand.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import xbot.common.properties.XPropertyManager;
+
+public class WinchExtendCommand extends BaseCommand{
+    WinchSubsystem winchSubsystem;
+    DoubleProperty winchExtentionPower;
+    
+    @Inject
+    public WinchExtendCommand (WinchSubsystem winchSubsystem, XPropertyManager propManager){
+        this.winchSubsystem = winchSubsystem;
+        winchExtentionPower = propManager.createPersistentProperty("winch extention power", -1.0);
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchExtendCommand.java]: requires

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchRetractCommand.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import xbot.common.properties.XPropertyManager;
+
+public class WinchRetractCommand extends BaseCommand{
+    WinchSubsystem winchSubsystem;
+    DoubleProperty winchRetractionPower;
+    
+    @Inject
+    public WinchRetractCommand (WinchSubsystem winchSubsystem, XPropertyManager propManager){
+        this.winchSubsystem = winchSubsystem;
+        winchRetractionPower = propManager.createPersistentProperty("winch retraction power", 1.0);
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchRetractCommand.java]: requires

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchStopCommand.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package competition.subsystems.hanger.winch_commands;
+
+import com.google.inject.Inject;
+
+import competition.subsystems.hanger.WinchSubsystem;
+import xbot.common.command.BaseCommand;
+
+public class WinchStopCommand extends BaseCommand{
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchStopCommand.java]: I'm assuming you want these stop commands to be the defaults on these subsystems?

**Comment by aschokking** [comment]
Should think of buttons on the operator joysticks to run these commands and add mappings for them

File: CompetitionTemplate/src/competition/subsystems/drive/commands/DriveToWallCommand.java [hunk 11]
```diff
@@ -44,8 +43,13 @@ public void execute() {
         drive.tankDriveSafely(power, power);
     }
     
+    public boolean isFinished(){
+        return Math.abs(targetDistanceFromWall - pose.getFrontRangefinderDistance()) < 1;
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/drive/commands/DriveToWallCommand.java]: Make this `1` a double property

File: CompetitionTemplate/src/competition/subsystems/hanger/HookSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import xbot.common.controls.actuators.XSpeedController;
+import xbot.common.injection.wpi_factories.WPIFactory;
+
+@Singleton
+public class HookSubsystem extends BaseSubsystem{
+    public XSpeedController hookMotor;
+    
+    @Inject
+    public HookSubsystem(WPIFactory factory) {
+        hookMotor = factory.getSpeedController(9);
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/HookSubsystem.java]: needs a logging statement like "Creating HookSubsystem"

**Comment by aschokking** [comment]
lgtm

File: CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import xbot.common.controls.actuators.XSpeedController;
+import xbot.common.injection.wpi_factories.WPIFactory;
+
+@Singleton
+public class WinchSubsystem extends BaseSubsystem{
+    public XSpeedController winchMotor;
+    
+    @Inject
+    public WinchSubsystem(WPIFactory factory) {
+        winchMotor = factory.getSpeedController(8);
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/WinchSubsystem.java]: needs a logging statement like "Creating WinchSubsystem"

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookExtendCommand.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import xbot.common.properties.XPropertyManager;
+
+public class HookExtendCommand extends BaseCommand{
+    public HookSubsystem hookSubsystem;
+    DoubleProperty hookExtentionPower;
+    
+    @Inject
+    public HookExtendCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
+        this.hookSubsystem = hookSubsystem;
+        hookExtentionPower = propManager.createPersistentProperty("hook extention power", -1.0);
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookExtendCommand.java]: To be in keeping with the rest of the robot - positive power should go up, and negative power should go down.

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookExtendCommand.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        this.requires(hookSubsystem);
+    }
+
+    @Override
+    public void initialize(){
+        hookSubsystem.setHookMotorPower(hookExtentionPower.get());
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookExtendCommand.java]: needs a logging statement like "Initializing HookExtendCommand"

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookRetractCommand.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    @Inject
+    public HookRetractCommand(HookSubsystem hookSubsystem, XPropertyManager propManager){
+        this.hookSubsystem = hookSubsystem;
+        hookRetractionPower = propManager.createPersistentProperty("hook retraction power", 1.0);
+        this.requires(hookSubsystem);
+    }
+
+    @Override
+    public void initialize(){
+        hookSubsystem.setHookMotorPower(hookRetractionPower.get());
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookRetractCommand.java]: needs a logging statement like "Initializing HookExtendCommand"

File: CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookStopCommand.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    
+    @Inject
+    public HookStopCommand(HookSubsystem hookSubsystem){
+        this.hookSubsystem =  hookSubsystem;
+        this.requires(hookSubsystem);
+    }
+
+    @Override
+    public void initialize() {
+        hookSubsystem.setHookMotorPower(0);
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/hook_commands/HookStopCommand.java]: Needs initializiation log

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchExtendCommand.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        this.requires(winchSubsystem);
+    }
+
+    @Override
+    public void initialize() {
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchExtendCommand.java]: Needs initializiation log

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchRetractCommand.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        this.requires(winchSubsystem);
+    }
+
+    @Override
+    public void initialize() {
+        winchSubsystem.setWinchMotorPower(winchRetractionPower.get());
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchRetractCommand.java]: Needs initializiation log

File: CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchStopCommand.java [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+    
+    @Inject
+    public WinchStopCommand(WinchSubsystem winchSubsystem){
+        this.winchSubsystem =  winchSubsystem;
+        this.requires(winchSubsystem);
+    }
+
+    @Override
+    public void initialize() {
+        winchSubsystem.setWinchMotorPower(0);
```
Comment (JohnGilb) [inline CompetitionTemplate/src/competition/subsystems/hanger/winch_commands/WinchStopCommand.java]: Needs initializiation log

**Pull request merged**